### ISSUE TITLE
FR: Sign in with Apple Support

### ISSUE THREAD

**Issue by ajkolean** [firebase/firebase-ios-sdk#3145]
Add support for authenticating with Sign in with Apple. Soon Apple will require that you implement this to be listed on the App Store if your App uses other 3rd party login providers.

https://developer.apple.com/sign-in-with-apple/
https://developer.apple.com/documentation/signinwithapplerestapi

**Comment by StevenEWright** [comment]
At least one of the places Apple is explicit about requiring Apple Sign In in the future: https://developer.apple.com/news/?id=06032019j

> Sign In with Apple will be available for beta testing this summer. It will be required as an option for users in apps that support third-party sign-in when it is commercially available later this year.

It's not clear if / how they will distinguish between AuthN and AuthZ use cases.

**Comment by Francescu** [comment]
Also, here is the WWDC video link:
https://developer.apple.com/videos/play/wwdc2019/706/

**Comment by Reedyuk** [comment]
It would also appear that Apple will eventually reject app store submissions with apps that contains SSO but not apples one.
Which is a little worrying.

**Comment by renkelvin** [comment]
@Reedyuk Thanks for the input. Is it officially documented?

**Comment by morganchen12** [comment]
@renkelvin The docs are here: https://developer.apple.com/documentation/authenticationservices

**Comment by renkelvin** [comment]
@morganchen12 I mean the source of "Apple will eventually reject app store submissions with apps that contains SSO but not apples one."

**Comment by andygeers** [comment]
It does say it in that blog post that StevenEWright shared at the very top of this thread

**Comment by renkelvin** [comment]
I think I misunderstood what @Reedyuk means by "SSO" above.
If it means "use sign in with Apple across apps", then we're well informed of it.
If it means "[enterprise single sign-on](https://developer.apple.com/documentation/authenticationservices?language=objc#topics)" introduced in iOS 13, I didn't see Apple has additional app store review guidelines over it.

**Comment by Reedyuk** [comment]
No, sorry for the confusion, i meant normal sso and not enterprise.
I think its a really harsh from Apple to make this mandatory and i think they even mention the apple sign on is to appear above all of the other sso buttons.

**Comment by AlexLike** [comment]
+1 Even though you can technically implement Apple Sign In with Firebase already (using Custom Auth), I'd really appreciate a native option with an integration in the Firebase Console.

**Comment by jacobg** [comment]
Here is the corresponding issue in the firebase-js-sdk in order to support hybrid apps (e.g., Cordova):

https://github.com/firebase/firebase-js-sdk/issues/1931

**Comment by renkelvin** [comment]
Thanks for the feature request. The Firebase Auth team is working on this.

**Comment by mackoj** [comment]
@renkelvin 
https://developer.apple.com/news/?id=06032019j

**Comment by sylvioleblanc** [comment]
Thanks @renkelvin ... Where’s the best place to listen for updates on the progress of the Firebase implementation of Sign-In with Apple?

**Comment by morganchen12** [comment]
On this issue. There's not much progress-wise that we can report in a meaningful way without implicitly creating unreliable timeline expectations, so there haven't been many updates here. However, we are committed to releasing Sign in with Apple support before Sign in with Apple is mandatory for new apps, which is expected to happen in fall of this year.

**Comment by Brian-Co** [comment]
Support for iOS 13 release would be much appreciated 🙏

**Comment by SloanHawthorne** [comment]
Any update on this?

**Comment by giladrom** [comment]
Same. This is going to be mandatory very soon. Any updates?

**Comment by mechmatt01** [comment]
Well, I doubt Apple will make you have it on launch. I may be wrong and although they are stern on their rules look at their support for the iPhone X mandate. It took almost a year for them to even support iMovie for the X. I am not hating on Apple, but I am saying that they usually don’t follow their deadlines, nor do they make others follow them until further down the road. But I could be wrong on this one.

**Comment by lidemin** [comment]
Hmm, the catch is if you want to please Apple which potentially features your app, you'd better do whatever they say...

**Comment by KrisConrad** [comment]
It’s also nice to support features that users want at launch. As a user, I’ve replaced apps that have taken too long to support functionality I thought was important. 

As a developer I don’t want to put my users in that position.

**Comment by mechmatt01** [comment]
Ok, points taken.

**Comment by sylvioleblanc** [comment]
I agree with Damon Lee.

Can we have a timeline for Support in Firebase Authentication?

Sent from my iPhone
________________________________

**Comment by triplef** [comment]
Cross-platforms iOS/Android apps will probably also want to offer this on Android for users switching from iOS, so I’ve opened a respective feature request for the Firebase Android SDK:

https://github.com/firebase/firebase-android-sdk/issues/700

**Comment by sylvioleblanc** [comment]
Definitely Frederik Seiffert!

Good call.

Sent from my iPhone
________________________________

**Comment by muinmomin** [comment]
While we're waiting on an official implementation from Firebase, is there a way to use "Sign in with Apple" with Firebase's custom authentication methods? Has anyone tested this out yet?

**Comment by SloanHawthorne** [comment]
If so how?

**Comment by StevenEWright** [comment]
@muinmomin @SloanHawthorne 

> While we're waiting on an official implementation from Firebase, has anyone successfully implemented "Sign in with Apple" with Firebase's custom authentication methods?

The problem is that a custom auth mechanism for Apple Sign In would require a server side component (to securely generate the required JWT) as well as the client side integration. The whole selling point of Firebase is that I don’t have to do those things.

**Comment by KrisConrad** [comment]
Unfortunately, being dependent on Firebase also means you’re going to have problems or be delayed implementing things like Login with Apple, Catalyst, Call Kit, or any number of RAM restricted app extensions Firebase is too resource heavy to run in until they decide it’s worth the effort to implement such things. 

There are always trade offs with these things.

**Comment by Brian-Co** [comment]
Well iOS 13 is less than 4 weeks away now... as is Sign-in with Apple being required...

**Comment by mackoj** [comment]
The issue with having dependency is that you can't never expect them to be ready on time with new mandatory or os feature...

**Comment by giladrom** [comment]
I'm actually quite disappointed this is still being "worked on". We should've had a beta two weeks after the announcement.

**Comment by mackoj** [comment]
@giladrom Auth0 has it working and testable June 17 https://auth0.com/blog/try-sign-in-with-apple-in-your-auth0-apps-today/

**Comment by buzzware** [comment]
My understanding is that if your app only offers traditional Firebase auth now, it won't be *required* to support Sign in with Apple because Firebase auth is not a third party - it would appear as a private identity. Offering Google account login however would qualify as third party and require Sign in with Apple too. Please correct me if I'm wrong.

**Comment by mackoj** [comment]
@buzzware I have the same understanding as you but the problem is if you use [Firebase Auth](https://firebase.google.com/products/auth/) service for authentication to Google, Facebook, Twitter.

**Comment by mackoj** [comment]
@westarle Thanks for the public comment and the ETA for this feature.

**Comment by akadebnath** [comment]
In case if anyone has missed it, here is update from App Store Review Guidelines:

> Sign In with Apple will be available for beta testing this summer. It will be required as an option for users in apps that support third-party sign-in when it is commercially available later this year.

It is present at the very bottom of this link:
https://developer.apple.com/news/?id=06032019j

I believe Apple will soon enforce including Sign in with Apple feature in the apps. So it is better if we can have it as early as possible. We, developers will surely need some time to integrate the feature in our apps, test the same and release and update.

**Comment by sipersso** [comment]
Seems pretty straightforward to build this using Firebase Cloud Functions and Custom Auth Tokens. Any downsides of using this in the meantime until official support from Firebase?

**Comment by JuhaniLehtimaeki** [comment]
The text does say that if your app exclusively use social media login you must include apple. If you enable phone number, email/passwd or anonymous you should be able to submit a new app even if you don't have an apple login.

**Comment by sceee** [comment]
@JuhaniLehtimaeki exactly, but my app uses exclusively Google Login until now.

**Comment by dahbiahmed** [comment]
> or anonymous
@JuhaniLehtimaeki do you have a source on the "anonymous" bit?

**Comment by pepejeria** [comment]
FWIW: I submitted an app with Xcode 11 GM that allows users to authenticate with Facebook, Gmail, Email or anonymously and it was approved.

**Comment by ksulliva** [comment]
> The text does say that if your app exclusively use social media login you must include apple. If you enable phone number, email/passwd or anonymous you should be able to submit a new app even if you don't have an apple login.

Unfortunately this is not true. The Apple guidelines are worded poorly, which makes this seem true, but basically if you use social auth for user accounts (regardless of whatever other methods are also used), you will need to support Sign in with Apple.

This was clarified in an update at the end of an article [here](https://techcrunch.com/2019/09/12/apple-tweaks-app-store-rule-changes-for-childrens-apps-and-sign-in-services/).

**Comment by cpboyd** [comment]
@LeXa777 It may seem simple for iOS, but I'm sure Firebase wants to enable it broadly across all Firebase platforms simultaneously.

It's potentially confusing with how they might map accounts.  A user can choose whether or not to use a random (or, essentially, anonymous) email.

Meanwhile, if the user gives their email address, it should probably be a "trusted" source (allowing it or Google login to be used interchangeably).

Contrast this behavior with "social" logins, like Facebook, where if you use Google sign-in, it prevents Facebook login from working since Google is "trusted."

Granted, if Google sticks with only their authentication being "trusted", then it may result with people simply dropping Google login support... Which is certainly something Google probably wants to avoid.

I, for one, want my users to be able to log-in (and create an account) as easily as possible, without having to remember whether or not they once clicked the "Google" button.

**Comment by giladrom** [comment]
Hi firebase team - any updates?

**Comment by Mayberries** [comment]
For now I found a work-around by using Firebase's authenticate with password/email. I ended up saving the credentials to keychain bc Sign in with Apple only provides the user's credentials with the first sign in. Hope this helps.

**Comment by cpboyd** [comment]
> Sign in with Apple only provides the user's credentials with the first sign in. Hope this helps.

@Mayberries Is that per-install? In other words, if a user deletes the app, will you get those same credentials when they sign-in after re-installing?

**Comment by sipersso** [comment]
@Mayberries @cpboyd Not sure that is a good idea. What are you using as a password? Apple can invalidate the token at any time. Not sure it will generate the same random email next time either. 

Someone can correct me if I am wrong, but I think singing in with a custom token is the only valid option until the firebase team has built this. A custom token can be generated by passing the apple sign in credentials to your server or a firebase cloud function that generates a token for you.

**Comment by cpboyd** [comment]
> Apple can invalidate the token at any time. Not sure it will generate the same random email next time either.

@sipersso That's my concern as well.

**Comment by Mayberries** [comment]
> > Sign in with Apple only provides the user's credentials with the first sign in. Hope this helps.
> 
> @Mayberries Is that per-install? In other words, if a user deletes the app, will you get those same credentials when they sign-in after re-installing?

Yes, bc the credentials are saved to keychain

**Comment by Mayberries** [comment]
> @Mayberries @cpboyd Not sure that is a good idea. What are you using as a password? Apple can invalidate the token at any time. Not sure it will generate the same random email next time either.
> 
> Someone can correct me if I am wrong, but I think singing in with a custom token is the only valid option until the firebase team has built this. A custom token can be generated by passing the apple sign in credentials to your server or a firebase cloud function that generates a token for you.

Just generating a random string for the pw. The privaterelay email generated by apple stays the same even after removing the app from the phone and removing the app from "Apps Using Your Apple ID".

**Comment by ramjyroo** [comment]
https://medium.com/@piazzachris/how-to-authenticate-a-user-with-sign-in-with-apple-and-firebase-using-a-custom-token-80dedbc24dd - seems to be a good starting for using google cloud functions as our custom backend to create the token

**Comment by bulby97** [comment]
@Mayberries may be cool for one device, but if there are more than one device, this will not work. I got lazy I used response.user as the password so that I can retrieve it on another device

**Comment by ramjyroo** [comment]
https://medium.com/@piazzachris/how-to-authenticate-a-user-with-sign-in-with-apple-and-firebase-using-a-custom-token-80dedbc24dd - seems to be a good starting for using google cloud functions as our custom backend to create the token

**Comment by Mayberries** [comment]
> @Mayberries may be cool for one device, but if there are more than one device, this will not work. I got lazy I used response.user as the password so that I can retrieve it on another device

I haven't tested it on multiple devices, but when I do I'll update you. I think it will depend on if you make the keychain synchronizable, and whether the private relay email that apple provides is generated for each device, or generated per AppleID (which makes more sense). If your keychain syncs across devices and the private relay email stays the same for your appleID then this method will work.

**Comment by audkar** [comment]
> I think it will depend on if you make the keychain synchronizable

How would you make keychain synchronizable for Firefox on Windows? e.g. user has iPhone and windows PC.

**Comment by Mayberries** [comment]
> > I think it will depend on if you make the keychain synchronizable
> 
> How would you make keychain synchronizable for Firefox on Windows? e.g. user has iPhone and windows PC.

In that case I think you would just have to look up your email/password in keychain. Again, this is just a work-around to satisfy apple's requirements.

**Comment by fishandphil** [comment]
Good evening
Now that iOS 13 has been out for "a while already"...
Is it possible to get an answer from anyone at Firebase regarding an official release? 
We are looking forward to Sign-in with Apple in our product but stuck. 
Visibility would make our life much easier :) 
Thank you. We love Firebase

**Comment by IchordeDionysos** [comment]
Please stop spamming!
As you noticed from the technical discussion in this thread it is far from trivial to implement Sign-in with Apple not mentioning any technical discussion on how to store/handling the data in the back end to make sure everything is secure!

And it's more than unusual to even get a timeline:

>Given the amount of work involved in adding a new custom IDP protocol, we are targeting release of native Apple Sign-In functionality on iOS in Q4 2019 barring any major technical hurdles.

**Comment by lukemcgregor** [comment]
The reason i'm subscribed to this thread is to get notifications when it's actually implemented (or wont be), not several a day from people who are upset its not done yet... 

If you want this feature thumbs up the issue rather than adding a new comment.

**Comment by LeXa777** [comment]
The reason I’m subscribed to this thread is not only to get a notification when this is implemented but also to get an idea when to expect it as I need to plan my time and my releases around this knowledge. I can’t just patiently wait until December when we get a happy post from Firebase team that they're are about to finally release the Alpha version.

I mean if they release it in December, it’s not great, but it’s not the end of the world as long as we know about this in advance and can make alternative plans (i.e. implement it our self, switch to Auth0, etc). This is a standard in software development to communicate expected dates (not necessary precise) to your stakeholders and users that rely on your software. 

Sorry for the spam.

**Comment by shliama** [comment]
Jeez. Everyone around here is so sensitive! No fun at all 😢
Not sure if it's spam 🤔Sorry Not Sorry.

**Comment by nicholausadi** [comment]
> Given the amount of work involved in adding a new custom IDP protocol, we are targeting release of native Apple Sign-In functionality on iOS in Q4 2019 barring any major technical hurdles.

Hi everyone, I just want to make sure is this still in track for Q4 2019? Or should we use this https://medium.com/@piazzachris/how-to-authenticate-a-user-with-sign-in-with-apple-and-firebase-using-a-custom-token-80dedbc24dd?

**Comment by morganchen12** [comment]
Sign in with Apple support is very close to release. Here's the [proposed changes to the quickstart](https://github.com/firebase/quickstart-ios/pull/808), where you can see what Sign in with Apple integration will look like.

Thanks everyone for your patience, and sorry for the lack of further transparency here.

**Comment by devcexx** [comment]
> > Given the amount of work involved in adding a new custom IDP protocol, we are targeting release of native Apple Sign-In functionality on iOS in Q4 2019 barring any major technical hurdles.
> 
> Hi everyone, I just want to make sure is this still in track for Q4 2019? Or should we use this https://medium.com/@piazzachris/how-to-authenticate-a-user-with-sign-in-with-apple-and-firebase-using-a-custom-token-80dedbc24dd?

I won't use that post "as-is" to implement Apple Sign In, since it seems that the author has forgotten to write the part where he actually verifies the authentication against Apple ID servers, from the server-side

**Comment by juanscolari** [comment]
Is there an estimated date when this feature will be released?

**Comment by LeXa777** [comment]
> Is there an estimated date when this feature will be released?

I doubt we will get any meaningful response, other then, 'we're working on it - should be out soon, blah blah'. I'm about to submit my app for review without the Apple Sign In button in a week or so and I'm a bit worried, but fingers crossed Apple is still flexible on this as otherwise I'm screwed.

**Comment by aplourde** [comment]
> > Is there an estimated date when this feature will be released?
> 
> I doubt we will get any meaningful response, other then, 'we're working on it - should be out soon, blah blah'. I'm about to submit my app for review without the Apple Sign In button in a week or so and I'm a bit worried, but fingers crossed Apple is still flexible on this as otherwise I'm screwed.

We've submitted 3 versions since the release of iOS 13 without Apple Sign In (we do offer Google, Facebook and usr/pwd authentication). Apple never complained about it yet !

**Comment by giladrom** [comment]
Same. I had Google, Facebook and Email logins enabled. I’ve since removed Google but still have Facebook and Email, Apple doesn’t care (for now).

**Comment by cpboyd** [comment]
> > I doubt we will get any meaningful response, other then, 'we're working on it - should be out soon, blah blah'. I'm about to submit my app for review without the Apple Sign In button in a week or so and I'm a bit worried, but fingers crossed Apple is still flexible on this as otherwise I'm screwed.
> 
> We've submitted 3 versions since the release of iOS 13 without Apple Sign In (we do offer Google, Facebook and usr/pwd authentication). Apple never complained about it yet !

FYI existing apps have a slightly longer period to add Apple Sign-in.

The exact wording makes it sound like **NEW** apps **must** have Apple Sign-in (if they offer any third-party sign-in services) if they want to be approved.

As I'm sure we're all well-aware, new apps typically go through a more stringent approval process than updates.

**Comment by LeXa777** [comment]
Great, thanks for confirming that folks, I guess I can be more relaxed now 🤠

**Comment by ksilz** [comment]
> FYI existing apps have a slightly longer period to add Apple Sign-in.
> 
> The exact wording makes it sound like **NEW** apps **must** have Apple Sign-in (if they offer any third-party sign-in services) if they want to be approved.

The [Apple guidelines](https://developer.apple.com/news/?id=09122019b) seem pretty clear:
- **New** apps must have Apple Sign-in now if they offer any third-party sign-in services.
- **Existing** apps (probably as in "in App Store already before September 12") with third-party sign-in service must have Apple Sign-in by April 2020 at the latest.

**Comment by giladrom** [comment]
If I were to submit a new app today, I’d start with E-Mail logins only and add social logins once this feature is available. Seems like the safest
option for now.

**Comment by Eldavid** [comment]
We just published an application with email/pwd and Facebook login without any problem. The guidle is clear: Apps that **exclusively**  use a third-party or social login service ...

**Comment by andreimateiro** [comment]
> We just published an application with email/pwd and Facebook login without any problem. The guidle is clear: Apps that **exclusively** use a third-party or social login service ... So if you have email login, U Will have no problem.

That's not actually true. I published a new app with only: social login + anonymous browse (no email/pw option)

**Comment by LeXa777** [comment]
Don't want to get into the guidelines debate here, just wanted to note however that Apple can be a bit lenient in the first few months or so when a new feature has been introduced (how lenient and for how long is anyone's guess). So although I find it really useful to hear other people experiences at **this moment** (as I'm about to submit my own app too in the next few days) I wouldn't make any long term conclusions based on this info alone.

**Comment by IchordeDionysos** [comment]
By the way found this comment on the Firebase UI ticket:

https://github.com/firebase/FirebaseUI-iOS/issues/698#issuecomment-541786694

**Comment by morganchen12** [comment]
This is available in the latest SDK release (6.13.0). See the [Sign in with Apple](https://firebase.google.com/docs/auth/ios/apple) guide for details on implementing it. If you encounter any bugs with the new Sign in with Apple behavior, please file them in a new issue.

Thanks all for your patience!

**Comment by knikolou** [comment]
Dear Unfortunately Apple start reject APPS because of Apple Sign In
Guideline 4.8 - Design - Sign in with Apple
We noticed that your app uses a third-party login service but does not offer Sign in with Apple.
Next Steps
To resolve this issue, please revise your app to offer Sign in with Apple as an equivalent login option.
Resources

**Comment by LouisCAD** [comment]
@knikolou Apple Sign In is available in Firebase in open beta, so you can use it straightaway.

**Comment by melsam** [comment]
When will Apple Sign In be added to FirebaseUI?

**Comment by morganchen12** [comment]
@melsam Sign in with Apple is available in [FirebaseUI 8.4.0](https://github.com/firebase/FirebaseUI-iOS/releases/tag/v8.4.0).

**Comment by ssbaval** [comment]
I can see Apple sign in (Beta) in Firestore console which is great. However, I cannot see any Flutter plugin that allows apple sign in ? Firebase Auth  0.15.0+1 doesnt have apple sign in as a provider - So how can we make use of this now ?

**Comment by paulb777** [comment]
@ssbaval Please check in the [Flutter community](https://github.com/flutter/flutter/issues).

**Comment by IAmKio** [comment]
Thanks for everyone's help with this - it's a great effort. I just wanted to add i've also started receiving app rejections because i do not offer Apple Sign In. Will be upgrading asap!

**Comment by rumnat** [comment]
Is it for new or existing apps?

**Comment by melsam** [comment]
> @melsam Sign in with Apple is available in [FirebaseUI 8.4.0](https://github.com/firebase/FirebaseUI-iOS/releases/tag/v8.4.0).

Where is the documentation for using Apple Sign In with FirebaseUI? None of the current FirebaseUI documentation mentions Apple. I tried adding pod 'FirebaseUI/Apple' to my podfile and that didn't work either.

**Comment by kplatfoot** [comment]
@melsam Add the pod `FirebaseUI/OAuth` and then include `FUIOAuth.appleAuthProvider()` in your auth provider list (only on iOS 13+, of course). Hope this helps!

**Comment by melsam** [comment]
> @melsam Add the pod `FirebaseUI/OAuth` and then include `FUIOAuth.appleAuthProvider()` in your auth provider list (only on iOS 13+, of course). Hope this helps!

Thanks @kplatfoot that was it!
